crates/tspoint/src/lib.rs

```diff
@@ -0,0 +1,161 @@
+
+use serde::{Deserialize, Serialize, ser::SerializeStruct};
+
+use flat_serialize_macro::FlatSerializable;
+
+use std::ffi::CStr;
+
+#[derive(Clone, Copy, PartialEq, Debug, FlatSerializable)]
+#[repr(C)]
+pub struct TSPoint {
+    pub ts: i64,
+    pub val: f64,
+}
+
+#[derive(Debug, PartialEq)]
+pub enum TSPointError {
+    TimesEqualInterpolate,
+}
+
+impl TSPoint {
+    pub fn interpolate_linear(&self, p2: &TSPoint, ts: i64) -> Result<f64, TSPointError> {
+        if self.ts == p2.ts {
+            return Err(TSPointError::TimesEqualInterpolate);
+        }
+        // using point slope form of a line iteratively y = y2 - y1 / (x2 - x1) * (x - x1) + y1
+        let duration = (p2.ts - self.ts) as f64; // x2 - x1
+        let dinterp = (ts - self.ts) as f64; // x - x1
+        Ok((p2.val - self.val) * dinterp / duration + self.val)
+    }
+}
+
+impl Serialize for TSPoint {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer {
+        if serializer.is_human_readable() {
+            // FIXME ugly hack to use postgres functions in an non-postgres library
+            extern "C" {
+                fn _ts_toolkit_encode_timestamptz(dt: i64, buf: &mut [u8; 128]);
+            }
+            let mut ts = [0; 128];
+            unsafe {
+                _ts_toolkit_encode_timestamptz(self.ts, &mut ts);
+            }
```

Comment:
(This code is _entirely_ my fault, and this comment is meant to cast aspersions on me, not Brian's wonderful PR)
I just realized we do might be able to do this more sanely by dynamically by storing a global function pointer (well, `AtomicUsize` but same diff), and setting it lazily from somewhere. This would also allow us to to have a default version of this function for testing outside of Postgres. It may pay for us to do this in a followup PR.

Reply:
Sounds promising, but we'll leave this for a future change.